# Constructing an Index Fund

In [1]:
import sys
!{sys.executable} -m pip install pulp

In [2]:
import pandas as pd
import xlrd
import openpyxl

df = pd.read_excel(r'EUROSTOXX50.xlsx', sheet_name='DailyReturns')

df.drop('Date',inplace=True,axis=1)
df

0         1         2         3         4         5         6   \
0          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1     0.000000  0.000000       NaN  0.000000  0.000000  0.000000  0.000000   
2     0.011242 -0.001969       NaN  0.005943 -0.002652 -0.001951  0.002793   
3          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
4          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
...        ...       ...       ...       ...       ...       ...       ...   
2186  0.005128  0.004809  0.001357  0.006746  0.001051  0.000000 -0.000814   
2187       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2188       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2189 -0.010070 -0.009231 -0.009217       NaN       NaN -0.009973       NaN   
2190 -0.013834       NaN  0.000000       NaN       NaN       NaN       NaN   

            7         8         9   ...        40        41        42  \
0          NaN       NaN       NaN  ...       NaN       NaN       NaN   
1     0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2    -0.000148  0.002488  0.007496  ...  0.018095  0.006911  0.013826   
3          NaN       NaN       NaN  ...       NaN       NaN       NaN   
4          NaN       NaN       NaN  ...       NaN       NaN       NaN   
...        ...       ...       ...  ...       ...       ...       ...   
2186  0.007550  0.005330 -0.005299  ... -0.007100  0.000400  0.002317   
2187       NaN       NaN       NaN  ...       NaN       NaN       NaN   
2188       NaN       NaN       NaN  ...       NaN       NaN       NaN   
2189 -0.014987 -0.008100 -0.005464  ...       NaN       NaN       NaN   
2190  0.003043       NaN       NaN  ...       NaN       NaN       NaN   

            43        44        45        46        47        48        49  
0          NaN       NaN       NaN       NaN       NaN       NaN       NaN  
1     0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
2     0.000649  0.005152  0.000803 -0.007658  0.003476  0.030626  0.000000  
3          NaN       NaN       NaN       NaN       NaN       NaN       NaN  
4          NaN       NaN       NaN       NaN       NaN       NaN       NaN  
...        ...       ...       ...       ...       ...       ...       ...  
2186  0.005300  0.006127  0.003254 -0.002359 -0.002693  0.010275  0.002838  
2187       NaN       NaN       NaN       NaN       NaN       NaN       NaN  
2188       NaN       NaN       NaN       NaN       NaN       NaN       NaN  
2189 -0.008896 -0.014378       NaN  0.002001       NaN -0.003736 -0.002377  
2190       NaN       NaN       NaN  0.003812       NaN       NaN       NaN  

[2191 rows x 50 columns]

In [3]:
corrMatrix=df.corr()

In [4]:
corrMatrix

0         1         2         3         4         5         6   \
0   1.000000  0.325478  0.141097  0.493510  0.406486  0.494116  0.376797   
1   0.325478  1.000000  0.276682  0.410769  0.390511  0.433510  0.384833   
2   0.141097  0.276682  1.000000  0.300074  0.370804  0.293699  0.465394   
3   0.493510  0.410769  0.300074  1.000000  0.549818  0.650188  0.439968   
4   0.406486  0.390511  0.370804  0.549818  1.000000  0.558609  0.440599   
5   0.494116  0.433510  0.293699  0.650188  0.558609  1.000000  0.477416   
6   0.376797  0.384833  0.465394  0.439968  0.440599  0.477416  1.000000   
7   0.384100  0.352610  0.312569  0.502602  0.457079  0.451298  0.452307   
8   0.482478  0.433862  0.235432  0.698272  0.535899  0.717191  0.456544   
9   0.506586  0.355034  0.216262  0.577448  0.459061  0.591412  0.407374   
10  0.449765  0.399782  0.147847  0.591836  0.528686  0.653124  0.417453   
11  0.544681  0.389178  0.191920  0.577115  0.461410  0.516444  0.430488   
12  0.405369  0.323379  0.219210  0.558985  0.506693  0.690470  0.392527   
13  0.437561  0.362734  0.288076  0.552962  0.538838  0.566629  0.453525   
14  0.457245  0.333220  0.242015  0.585908  0.544537  0.783244  0.423582   
15  0.472232  0.407683  0.200176  0.605325  0.561874  0.683484  0.431148   
16  0.396641  0.356004  0.239441  0.485033  0.421197  0.568802  0.425888   
17  0.482634  0.402775  0.345767  0.578220  0.522045  0.566443  0.467445   
18  0.437425  0.408784  0.242764  0.576236  0.527999  0.651096  0.475719   
19  0.520470  0.379198  0.126386  0.609666  0.455776  0.631423  0.423140   
20  0.460985  0.373122  0.231542  0.543385  0.457755  0.512264  0.408034   
21  0.479568  0.320752  0.147176  0.548303  0.435172  0.556279  0.398998   
22  0.468691  0.321846  0.249870  0.547032  0.436827  0.534969  0.405008   
23  0.439257  0.312195  0.303155  0.583656  0.466734  0.599830  0.390763   
24  0.225418  0.181962  0.178356  0.221849  0.194355  0.223302  0.230505   
25  0.425097  0.319410  0.308202  0.580413  0.468433  0.582505  0.386352   
26  0.445026  0.305871  0.078230  0.497588  0.366025  0.501425  0.390553   
27  0.360730  0.371305  0.228218  0.474164  0.434771  0.507247  0.457619   
28  0.418603  0.322105  0.236755  0.544519  0.484641  0.680974  0.408696   
29  0.368818  0.239082  0.213076  0.493449  0.437582  0.624024  0.355162   
30  0.504586  0.395430  0.259459  0.537267  0.453879  0.553256  0.459480   
31  0.377428  0.392429  0.423173  0.523056  0.466733  0.489405  0.438393   
32  0.351999  0.325010  0.247537  0.456007  0.395778  0.452126  0.376554   
33  0.191480  0.189338  0.248389  0.356606  0.289464  0.315909  0.259352   
34  0.461621  0.469645  0.369766  0.623827  0.542048  0.580881  0.467346   
35  0.440808  0.356611  0.280779  0.588893  0.488729  0.780381  0.422426   
36  0.550389  0.419728  0.274101  0.628272  0.479778  0.548336  0.467744   
37  0.439329  0.382235  0.265004  0.573234  0.531472  0.558152  0.449899   
38  0.230729  0.323745  0.178391  0.276691  0.265318  0.478377  0.234829   
39  0.507695  0.361997  0.248408  0.530511  0.417335  0.443654  0.411434   
40  0.389311  0.348465  0.359450  0.556042  0.655089  0.531736  0.458004   
41  0.419214  0.292310  0.270205  0.540631  0.481602  0.662587  0.372270   
42  0.468896  0.336070  0.187220  0.546165  0.434741  0.529246  0.369316   
43  0.452028  0.420203  0.342182  0.584248  0.495945  0.591843  0.556486   
44  0.452346  0.416496  0.281481  0.640508  0.550712  0.677761  0.459714   
45  0.451861  0.399204  0.317257  0.665558  0.579555  0.649343  0.458809   
46  0.237939  0.219447  0.220993  0.391778  0.290716  0.405146  0.248350   
47  0.396774  0.297918  0.308292  0.491503  0.446734  0.519714  0.369303   
48  0.330092  0.243603  0.106221  0.330807  0.251726  0.323320  0.304834   
49  0.361405  0.300084  0.196745  0.455504  0.425657  0.542945  0.344571   

          7         8         9   ...        40        41        42        43  \
0   0.384100  0.482478  0.506586  ...  0.389311  0.4

In [5]:
from pulp import LpMaximize, LpProblem, LpStatus, LpVariable, lpSum

In [6]:
#some helper functions
def doubletosingle(i,j,n=50):
    return n*i+j

def singletodouble(x,n=50):
    b = x%n
    a = (x-b)//n
    return(a,b)

In [7]:
#For each stock i=0,..,49 the variable x_ij = 1 if j is the most similar stock in the index fund, 0 otherwise)
x = {k: LpVariable(name=f"x{k}", lowBound=0, upBound=1, cat="Integer") for k in range(2500)}

#The varibles y_j describe which stocks j are in the index fund (y_j=1 if j is selected in the fund, 0 otherwise)
y = {j: LpVariable(name=f"y{j}", lowBound=0, upBound=1, cat="Integer") for j in range(50)}

In [8]:
ll = [corrMatrix[i][j]*x[doubletosingle(i,j)] for i in range(50) for j in range (50)]

In [9]:
model = LpProblem(name="Fund", sense=LpMaximize)

#Objective function: maximises the similarity between the n stocks and their representatives in the fund.
model += lpSum(ll)

In [10]:
#Constraint 1: there must be 10 stocks in the fund
model += lpSum(y[j] for j in range(50)) == 10

In [11]:
#Constraint 2: Each stock i has exactly one representative stock j in the fund.
for i in range(50):
    model += lpSum(x[doubletosingle(i,j)] for j in range(50)) ==1

In [12]:
#Constraint 3: stock i can be represented by stock j only if j is in the fund.
for i in range(50):
    for j in range(50):
        model += y[j] >= x[doubletosingle(i,j)]

status = model.solve()

In [13]:
print(f"{LpStatus[model.status]}")
print(f"objective: {model.objective.value()}")

zz = []
for var in x.values():
    if var.value()>0:
        print(f"{var.name}: {var.value()}")
        zz.append(int(var.name[1:]))

Optimal
objective: 35.26565086041226
x36: 1.0
x55: 1.0
x102: 1.0
x155: 1.0
x215: 1.0
x255: 1.0
x305: 1.0
x365: 1.0
x415: 1.0
x465: 1.0
x515: 1.0
x586: 1.0
x612: 1.0
x665: 1.0
x705: 1.0
x765: 1.0
x805: 1.0
x871: 1.0
x905: 1.0
x955: 1.0
x1036: 1.0
x1071: 1.0
x1121: 1.0
x1175: 1.0
x1224: 1.0
x1275: 1.0
x1321: 1.0
x1365: 1.0
x1412: 1.0
x1462: 1.0
x1505: 1.0
x1565: 1.0
x1605: 1.0
x1683: 1.0
x1736: 1.0
x1755: 1.0
x1836: 1.0
x1855: 1.0
x1912: 1.0
x1986: 1.0
x2005: 1.0
x2062: 1.0
x2136: 1.0
x2155: 1.0
x2215: 1.0
x2255: 1.0
x2325: 1.0
x2355: 1.0
x2448: 1.0
x2465: 1.0


In [14]:
zz

[36,
 55,
 102,
 155,
 215,
 255,
 305,
 365,
 415,
 465,
 515,
 586,
 612,
 665,
 705,
 765,
 805,
 871,
 905,
 955,
 1036,
 1071,
 1121,
 1175,
 1224,
 1275,
 1321,
 1365,
 1412,
 1462,
 1505,
 1565,
 1605,
 1683,
 1736,
 1755,
 1836,
 1855,
 1912,
 1986,
 2005,
 2062,
 2136,
 2155,
 2215,
 2255,
 2325,
 2355,
 2448,
 2465]

In [15]:
yy=[]
for var in y.values():
    if var.value()>0:
        print(f"{var.name}: {var.value()}")
        yy.append(int(var.name[1:]))

y2: 1.0
y5: 1.0
y12: 1.0
y15: 1.0
y21: 1.0
y24: 1.0
y25: 1.0
y33: 1.0
y36: 1.0
y48: 1.0


Stock 3,6,13,16,22,25,26,34,37 and 49 are chosen to be in the index fund

In [16]:
yy

[2, 5, 12, 15, 21, 24, 25, 33, 36, 48]

In [17]:
singletodouble(36), singletodouble(55), singletodouble(102), singletodouble(155), singletodouble(215), singletodouble(255)

((0, 36), (1, 5), (2, 2), (3, 5), (4, 15), (5, 5))

In [18]:
singletodouble(305), singletodouble(365), singletodouble(415), singletodouble(465), singletodouble(515), singletodouble(586)

((6, 5), (7, 15), (8, 15), (9, 15), (10, 15), (11, 36))

In [19]:
singletodouble(612), singletodouble(665), singletodouble(705), singletodouble(765), singletodouble(805), singletodouble(871)

((12, 12), (13, 15), (14, 5), (15, 15), (16, 5), (17, 21))

In [20]:
singletodouble(905), singletodouble(955), singletodouble(1036), singletodouble(1071), singletodouble(1121)

((18, 5), (19, 5), (20, 36), (21, 21), (22, 21))

In [21]:
singletodouble(1175), singletodouble(1224), singletodouble(1275), singletodouble(1321), singletodouble(1365)

((23, 25), (24, 24), (25, 25), (26, 21), (27, 15))

In [22]:
singletodouble(1412), singletodouble(1462), singletodouble(1505), singletodouble(1565), singletodouble(1605)

((28, 12), (29, 12), (30, 5), (31, 15), (32, 5))

In [23]:
singletodouble(1683), singletodouble(1736), singletodouble(1755), singletodouble(1836), singletodouble(1855), singletodouble(1912)

((33, 33), (34, 36), (35, 5), (36, 36), (37, 5), (38, 12))

In [24]:
singletodouble(1986), singletodouble(2005), singletodouble(2062), singletodouble(2136), singletodouble(2155)

((39, 36), (40, 5), (41, 12), (42, 36), (43, 5))

In [25]:
singletodouble(2215), singletodouble(2255), singletodouble(2325), singletodouble(2355), singletodouble(2448), singletodouble(2465)

((44, 15), (45, 5), (46, 25), (47, 5), (48, 48), (49, 15))

Stock 3 represents itself

Stock 6 represents Stock 2,4,7,15,17,19,20,31,33,36,38,41,44,46,48 and itself

Stock 13 represents Stock 29,30,39,42 and itself

Stock 16 represents Stock 5,8,9,10,11,14,28,32,45,50 and itself

Stock 22 represents Stock 18,23,27 and itself

Stock 25 represents itself

Stock 26 represents Stock 24,47 and itself

Stock 34 represents itself

Stock 37 represents Stock 1,12,21,35,40,43 and itself

Stock 49 represents itself

In [26]:
df2 = pd.read_excel(r'EUROSTOXX50.xlsx', sheet_name='MarketValue')
df2

StockNo  Market Value StockName  2019-12-02 00:00:00
0         0     67.005058    ABI.BR                  NaN
1         1    269.626465    ADS.DE                  NaN
2         2    683.000000  ADYEN.AS                  NaN
3         3    116.906754     AI.PA                  NaN
4         4    125.667252    AIR.PA                  NaN
5         5    200.849716    ALV.DE                  NaN
6         6     70.890541    AMS.MC                  NaN
7         7    236.541901   ASML.AS                  NaN
8         8     58.965755    BAS.DE                  NaN
9         9     64.558289   BAYN.DE                  NaN
10       10     68.091499    BMW.DE                  NaN
11       11     68.417809     BN.PA                  NaN
12       12     44.849205    BNP.PA                  NaN
13       13     32.508785    CRG.IR                  NaN
14       14     21.417801     CS.PA                  NaN
15       15     46.933178    DAI.DE                  NaN
16       16    133.026443    DB1.DE                  NaN
17       17     93.849617     DG.PA                  NaN
18       18     30.817467    DPW.DE                  NaN
19       19     13.605693    DTE.DE                  NaN
20       20    135.178345     EL.PA                  NaN
21       21      6.214023   ENEL.MI                  NaN
22       22     12.860342   ENGI.PA                  NaN
23       23     12.677710    ENI.MI                  NaN
24       24    100.681183   FLTR.IR                  NaN
25       25     42.069656     FP.PA                  NaN
26       26      8.140998    IBE.MC                  NaN
27       27     18.715788    IFX.DE                  NaN
28       28      9.329022   INGA.AS                  NaN
29       29      1.955902    ISP.MI                  NaN
30       30     27.226368    ITX.MC                  NaN
31       31    520.660950    KER.PA                  NaN
32       32     53.614162  KNEBV.HE                  NaN
33       33    196.426773       LIN                  NaN
34       34    385.207916     MC.PA                  NaN
35       35    244.441422   MUV2.DE                  NaN
36       36    244.145370     OR.PA                  NaN
37       37     38.698845   PHIA.AS                  NaN
38       38     60.517864    PRX.AS                  NaN
39       39    159.332504     RI.PA                  NaN
40       40    144.850006    SAF.PA                  NaN
41       41      3.176049    SAN.MC                  NaN
42       42     79.605568    SAN.PA                  NaN
43       43    115.985542    SAP.DE                  NaN
44       44    109.327057    SIE.DE                  NaN
45       45     83.761299    SU.PA1                  NaN
46       46     46.648590       TOT                  NaN
47       47     23.853888    VIV.PA                  NaN
48       48     43.356728    VNA.DE                  NaN
49       49    168.597824   VOW3.DE                  NaN

In [29]:
Total = df2['Market Value'].sum()

In [28]:
Total

5544.786922

In [30]:
index_weights = {x:0.0 for x in yy}
for l in zz:
    stock,rep = singletodouble(l)
    index_weights[rep] += df2['Market Value'][stock]/Total

In [31]:
index_weights

{2: 0.12317876405494811,
 5: 0.2866787190131812,
 12: 0.021610937207444236,
 15: 0.2616093816418073,
 21: 0.021834018457887277,
 24: 0.018157809202825847,
 25: 0.018286718214128694,
 33: 0.03542548627443902,
 36: 0.20539879819028325,
 48: 0.007819367743054995}

The proportion we should invest in Stock 3 is 0.1232

The proportion we should invest in Stock 6 is 0.2867

The proportion we should invest in Stock 13 is 0.0216

The proportion we should invest in Stock 16 is 0.2616

The proportion we should invest in Stock 22 is 0.0218

The proportion we should invest in Stock 25 is 0.0182

The proportion we should invest in stock 26 is 0.0183

The proportion we should invest in stock 34 is 0.03543

The proportion we should invest in stock 37 is 0.2054

The proportion we should invest in stock 49 is 0.0078